src/AppInstallerRepositoryCore/PackageDependenciesValidation.cpp

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+#include "pch.h"
+#include <AppInstallerVersions.h>
+#include <winget/Manifest.h>
+#include <winget/ManifestValidation.h>
+#include <SQLiteWrapper.h>
+#include <PackageDependenciesValidation.h>
+#include <Microsoft/Schema/1_4/DependenciesTable.h>
+#include "Microsoft/Schema/1_0/ManifestTable.h"
+#include <winget/DependenciesGraph.h>
+
+namespace AppInstaller::Repository
+{
+	using namespace Microsoft::Schema::V1_4;
+	using namespace Microsoft::Schema::V1_0;
+
+	namespace
+	{
+		std::vector<AppInstaller::Manifest::Dependency> GetDependencies(
+			const Manifest::Manifest& manifest, AppInstaller::Manifest::DependencyType dependencyType)
+		{
+			std::vector<AppInstaller::Manifest::Dependency>  dependencies;
+
+			for (const auto& installer : manifest.Installers)
+			{
+				installer.Dependencies.ApplyToAll([&](AppInstaller::Manifest::Dependency dependency)
+					{
+						if (dependency.Type == dependencyType)
+						{
+							dependencies.emplace_back(dependency);
+						}
+					});
+			}
+
+			return dependencies;
+		}
+	};
+
+	bool PackageDependenciesValidation::ValidateManifestDependencies(SQLiteIndex* index, const Manifest::Manifest manifest)
+	{
+		using namespace Manifest;
+
+		Dependency rootId(DependencyType::Package, manifest.Id, manifest.Version);
+		std::vector<ValidationError> dependenciesError;
+		bool foundErrors = false;
+
+		DependencyGraph graph(rootId, [&](const Dependency& node) {
+
+			DependencyList depList;
+			if (node.Id == rootId.Id)
```

Comment:
The special case is needed, it can be handled in the lambda function or outside(before the dependencies graph initialization),  it just makes sense to have the logic to handle dependencies evaluation in single place irrespective of whether it is root or child dependency .

Reply:
Because it is only checking for `Id` equality, a loop in the graph could get back to this special case.  That is itself very strange and I'm not even sure what we should do in that case at the moment.  If you are adding the highest version, then it is correct.  If you are not, then we have a strange situation.

Reply:
The dependency graph never process a node twice(so loops are not an issue) so the lambda is never called to the same node. Also, we're adding the highest version.